Remove unused STC.tls

Geod24 opened the pull request:
```
STC.tls was originally set on __thread variables.
However, __thread was removed from the language in 2013
(20ef96fe1d5c1631f5aead01873f17ca85cc944b),
as TLS was made the default for global variables.
The STC was however left in place, but was only ever set
in one place: for the gate in [shared] static {c,d}tors.
However, this was likely to be an accident, and it also looks wrong:
the gate for shared static {c,d}tor was static (which is TLS),
while the gate for non-shared module {c,d}tors was STC.tls,
which had no effect (so it was effectively a local variable).

However, that bug was not visible, thanks to the glue layer,
which has special handling of gate variables, essentially
ignoring the STC.
```

Geod24 committed:
Remove unused STC.tls

STC.tls was originally set on __thread variables.
However, __thread was removed from the language in 2013
(20ef96fe1d5c1631f5aead01873f17ca85cc944b),
as TLS was made the default for global variables.
The STC was however left in place, but was only ever set
in one place: for the gate in [shared] static {c,d}tors.

However, this was likely to be an accident, and it also looks wrong:
the gate for shared static {c,d}tor was static (which is TLS),
while the gate for non-shared module {c,d}tors was STC.tls,
which had no effect (so it was effectively a local variable).

However, that bug was not visible, thanks to the glue layer,
which has special handling of gate variables, essentially
ignoring the STC.

Geod24 commented:
```
../.dub/packages/eventcore-0.9.7/eventcore/source/eventcore/drivers/posix/processes.d: Error: variable `eventcore.drivers.posix.processes.PosixEventDriverProcesses!(EpollEventLoop).PosixEventDriverProcesses._sharedStaticCtor_L306_C2_1.__gate` __gshared not allowed in safe functions; use shared
```

Interesting, this is not covered by the test-suite. And the message could use improvement. I changed it to `shared`, let's see if the CI prefers this. Will raise another PR for this improvement.

UplinkCoder reviewed (approved):
Nice simplification.

UplinkCoder commented:
Error: read-modify-write operations are not allowed for `shared` variables
       Use `core.atomic.atomicOp!"+="(__gate, 1)` instead
runnable/xtest46_gc.d-mixin-37(4699): Error: template instance
`xtest46_gc.Hoge6691!()` error instantiating


Am Do., 6. Aug. 2020 um 23:22 Uhr schrieb Walter Bright <
notifications@github.com>:

> *@WalterBright* approved this pull request.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/dlang/dmd/pull/11518#pullrequestreview-462886076>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVWSCE5EZU4IFCDCO7NPJLR7MNKFANCNFSM4PWDWSMQ>
> .
>

Geod24 commented:
I added a teeny tiny hack to work around the issue and documented in, plus created an issue for the larger bug, which is probably a good starter issue: https://issues.dlang.org/show_bug.cgi?id=21147

MoonlightSentinel commented:
This should fix the incorrect module ctor behaviour in `runnable/test18868.d`:

```diff
diff --git a/src/dmd/dsymbolsem.d b/src/dmd/dsymbolsem.d
index a8d4b806e..3b2b8c575 100644
--- a/src/dmd/dsymbolsem.d
+++ b/src/dmd/dsymbolsem.d
@@ -4354,8 +4354,8 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
              * during static construction.
              */
             auto v = new VarDeclaration(Loc.initial, Type.tint32, Id.gate, null);
-            v.storage_class = STC.temp |
-                (scd.isSharedStaticCtorDeclaration() ? STC.shared_ : STC.static_);
+            v.storage_class = STC.temp | STC.static_ |
+                (scd.isSharedStaticCtorDeclaration() ? STC.shared_ : 0);
 
             auto sa = new Statements();
             Statement s = new ExpStatement(Loc.initial, v);
@@ -4432,8 +4432,8 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
              * during static destruction.
              */
             auto v = new VarDeclaration(Loc.initial, Type.tint32, Id.gate, null);
-            v.storage_class = STC.temp |
-                (sdd.isSharedStaticDtorDeclaration() ? STC.shared_ : STC.static_);
+            v.storage_class = STC.temp | STC.static_ |
+                (sdd.isSharedStaticDtorDeclaration() ? STC.shared_ : 0);
 
             auto sa = new Statements();
             Statement s = new ExpStatement(Loc.initial, v);
```

Allthough it seems to pass without the second line (which is weird if I understand this code correctly):
```
(scd.isSharedStaticCtorDeclaration() ? STC.shared_ : 0);
```

MoonlightSentinel commented on src/dmd/astbase.d:
```diff
@@ -109,7 +109,7 @@ struct ASTBase
         nothrow_            = (1L << 25),   // never throws exceptions
         pure_               = (1L << 26),   // pure function
-        tls                 = (1L << 27),   // thread local
+        //                    (1L << 27),
```
Remove the other declaration as well:

https://github.com/dlang/dmd/blob/fe263f5991ea571b0384c6facfe207a7b17efb44/src/dmd/declaration.d#L229

Geod24 committed:
Use atomicOp for increment/decrement in templated static ctor|dtor

The code previously relied on STC.tls to bypass shared RMW checks.
Removing STC.tls, which has no other uses, exposes the issue.
Instead of building another exception within the language,
using core.atomic seems cleaner, and the performance implications
are negligeable or inexistant as this is a very rare use-case
(not to mention that templated module ctor|dtor is currently broken
in separate compilation).

12345swordy commented:
What is the hold up on here?

Geod24 commented:
I originally assumed it was a trivial exercise and was proven wrong. So this PR turned into a high effort / low reward deal, which is why it was stuck for so long. Came back to it recently and hopefully the first commit is the right fix, let's see what the CI says.
